router/h2/src/e2e/scala/io/buoyant/router/h2/RouterEndToEndTest.scala

```diff
@@ -44,4 +44,132 @@ class RouterEndToEndTest
+      try {
+        await(client.close())
+        await(dog.server.close())
+        await(router.close())
+      } catch {
+        case e: Throwable => log.error(e, "closing")
+      } finally setLogLevel(Level.OFF)
+    }
+  }
+
+  test("resets downstream on upstream disconnect") {
+    val dogReqP, dogRspP = new Promise[Stream]
+    val dog = Downstream.service("dog") { req =>
+      dogReqP.setValue(req.stream)
+      dogRspP.map(Response(Status.Ok, _))
+    }
+
+    val dtab = Dtab.read(s"""
+        /p => /$$/inet/127.1 ;
+        /h2/clifford => /p/${dog.port} ;
+      """)
+    val identifierParam = H2.Identifier { _ =>
+      req => {
+        val dst = Dst.Path(Path.Utf8("h2", req.authority), dtab)
+        Future.value(new RoutingFactory.IdentifiedRequest(dst, req))
+      }
+    }
+    val router = H2.serve(new InetSocketAddress(0), H2.router
+      .configured(identifierParam)
+      .factory())
+    val client = upstream(router)
+    try {
+      val clientLocalStream, serverLocalStream = Stream()
+      val req = Request("http", Method.Get, "clifford", "/path", clientLocalStream)
+      val rspF = client(req)
+      val reqStream = await(dogReqP)
+      assert(!rspF.isDefined)
+
+      dogRspP.setValue(serverLocalStream)
+      val rsp = await(rspF)
+      val reqReadF = reqStream.read()
+      val rspReadF = rsp.stream.read()
+
+      await(client.close())
+      assert(await(reqReadF.liftToTry) == Throw(Reset.Cancel))
+      assert(await(rspReadF.liftToTry) == Throw(Reset.Cancel))
+
+    } finally {
+      setLogLevel(Level.OFF)
+      try await(dog.server.close().before(router.close()))
+      catch { case e: Throwable => log.error(e, "closing") }
+    }
+  }
+
+  test("resets upstream on downstream failure") {
+    val dogReqP, dogRspP = new Promise[Stream]
+    val dog = Downstream.service("dog") { req =>
+      dogReqP.setValue(req.stream)
+      dogRspP.map(Response(Status.Ok, _))
+    }
+
+    val dtab = Dtab.read(s"""
+        /p => /$$/inet/127.1 ;
+        /h2/clifford => /p/${dog.port} ;
+      """)
+    val identifierParam = H2.Identifier { _ =>
+      req => {
+        val dst = Dst.Path(Path.Utf8("h2", req.authority), dtab)
+        Future.value(new RoutingFactory.IdentifiedRequest(dst, req))
+      }
+    }
+    val router = H2.serve(new InetSocketAddress(0), H2.router
+      .configured(identifierParam)
+      .factory())
+    val client = upstream(router)
+    try {
+      val clientLocalStream = Stream()
+      val req = Request("http", Method.Get, "clifford", "/path", clientLocalStream)
+      val rspF = client(req)
+      val _ = await(dogReqP)
+      assert(!rspF.isDefined)
+
+      dogRspP.setException(Failure("failz").flagged(Failure.Rejected))
+      assert(await(rspF.liftToTry) == Throw(Reset.Refused))
+
+    } finally {
+      try await(client.close().before(dog.server.close()).before(router.close()))
+      catch { case e: Throwable => log.error(e, "closing") }
+      finally setLogLevel(Level.OFF)
```

Comment:
I'd hope so, but honestly the logging stuff gets wonky _all_ the time, so i've taken to being OCD about clearing it out